index/postings.go

```diff
@@ -117,9 +122,61 @@ func (p *memPostings) ensureOrder() {
 	p.ordered = true
 }
 
-// add adds a document to the index. The caller has to ensure that no
-// term argument appears twice.
-func (p *memPostings) add(id uint64, lset labels.Labels) {
+// Delete removes all ids in the given map from the postings lists.
+func (p *MemPostings) Delete(deleted map[uint64]struct{}) {
+	var keys []labels.Label
+
+	p.mtx.RLock()
+	for l := range p.m {
+		keys = append(keys, l)
+	}
+	p.mtx.RUnlock()
+
+	for _, l := range keys {
+		p.mtx.Lock()
```

Comment:
We are locking and unlocking repeatedly here, any reason why we are not doing a lock before and after the loop?

Reply:
Simply for locking granularity. Walking every single posting and processing deletes on it may well take a bit of time. If we locked until we have processed each, we'd potentially block any kind of read for multiple seconds.

Reply:
Could you also include this as a comment? Will help anyone reading the code.